### PR TITLE
add v0 view for program modification, tests

### PR DESCRIPTION
**Pull request by jansenk**
Added v0 mock view for PATCH requests to update program enrollment

### CODE REVIEW

File: registrar/apps/api/v0/tests/test_views.py [hunk 1]
```diff
@@ -266,3 +282,85 @@ def test_enrollment_payload_limit(self):
...
+    def test_200_ok(self):
+        patch_data = [
+            self.learner_modification("A", "enrolled"),
+            self.learner_modification("B", "pending"),
+            self.learner_modification("C", "suspended"),
+            self.learner_modification("D", "canceled"),
+        ]
+        response = self.patch(self.path, patch_data, self.user)
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(
```
Comment (kdmccormick) [inline registrar/apps/api/v0/tests/test_views.py]: I recommend `self.assertDictEqual` here. Will test the same thing as `self.assertEqual`, but gives more helpful errors when the assertion fails.

File: registrar/apps/api/v0/tests/test_views.py [hunk 2]
```diff
@@ -266,3 +282,85 @@ def test_enrollment_payload_limit(self):
...
+        path_404 = 'programs/nonexistant-program-will-404/enrollments/'
+        patch_data = [self.learner_modification("A", "enrolled")]
+        response = self.patch(path_404, patch_data, self.user)
+        self.assertEqual(404, response.status_code)
+
+    def test_422_unprocessable_entity(self):
+        patch_data = [{'student_id': 'A', 'status': 'this-is-not-a-status'}]
+        response = self.patch(self.path, patch_data, self.user)
+        self.assertEqual(422, response.status_code)
+        self.assertEqual({'A': 'invalid_status'}, response.data)
```
Comment (kdmccormick) [inline registrar/apps/api/v0/tests/test_views.py]: `self.assertDictEqual`

File: registrar/apps/api/v0/tests/test_views.py [hunk 3]
```diff
@@ -266,3 +282,85 @@ def test_enrollment_payload_limit(self):
...
+        """ Also tests duplicates """
+        patch_data = [
+            self.learner_modification("A", "enrolled"),
+            self.learner_modification("A", "enrolled"),
+            self.learner_modification("B", "not-a-status"),
+            self.learner_modification("C", "enrolled"),
+        ]
+        response = self.patch(self.path, patch_data, self.user)
+        self.assertEqual(response.status_code, 207)
+        self.assertEqual(
```
Comment (kdmccormick) [inline registrar/apps/api/v0/tests/test_views.py]: `self.assertDictEqual`

File: registrar/apps/api/v0/views.py [hunk 4]
```diff
@@ -195,3 +196,50 @@ def post(self, request, *args, **kwargs):
             return Response(response, HTTP_207_MULTI_STATUS)
         else:
             return Response(response)
+
+    def patch(self, request, *args, **kwargs):
```
Comment (zacharis278) [inline registrar/apps/api/v0/views.py]: Can we break some of this out into a shared function?  It looks to me like POST/PATCH validations are doing the same thing except for the serializer validation.

**Review by zacharis278 — CHANGES_REQUESTED** [review]

**Review by kdmccormick — COMMENTED** [review]
Looking good. Only comments are nits. Ping me when you rebase and I'll give a thumbs up.

File: registrar/apps/api/v0/views.py [hunk 5]
```diff
@@ -195,3 +196,50 @@ def post(self, request, *args, **kwargs):
...
+            raise ValidationError()
+        if len(request.data) > 25:
+            message = "Only 25 students may be specified at a time"
+            return Response(message, status=HTTP_413_REQUEST_ENTITY_TOO_LARGE)
+
+        response = {}
+        seen_ids = set()
+        good_status = bad_status = False
+        for learner_modification in request.data:
+            if not isinstance(learner_modification, dict):
```
Comment (iloveagent57) [inline registrar/apps/api/v0/views.py]: Nit: I think you could squish this down to fewer return statements, or possibly extract it to a helper function, something like:
```python
validations = {
    'record must be a dict': lambda: isinstance(learner_modification, dict),
    'student_id required': lambda: 'student_id' in learner_modification,
    'status required': lambda: 'status required' in learner_modification,
}
for msg, validation_func in validations.items():
    if not validation_func():
        message = 'malformed request data, {}'.format(msg)
        return Response(message, 422)
```

File: registrar/apps/api/v0/views.py (hunk 4 above)
Comment (kdmccormick) [reply registrar/apps/api/v0/views.py]: Looking at the methods again, I think we could use entirely the same method, just with a different serializer.

File: registrar/apps/api/v0/views.py [hunk 6]
```diff
@@ -195,3 +196,50 @@ def post(self, request, *args, **kwargs):
...
+                message = 'malformed request data, status required'
+                return Response(message, status=HTTP_422_UNPROCESSABLE_ENTITY)
+
+            student_id = learner_modification['student_id']
+            request_status = learner_modification['status']
+
+            if student_id in seen_ids:
+                response[student_id] = 'duplicated'
+                bad_status = True
+            elif request_status not in ['enrolled', 'pending', 'suspended', 'canceled']:
```
Comment (iloveagent57) [inline registrar/apps/api/v0/views.py]: Nit: it's probably useful to have this list of allowed statuses extracted into a global variable.

File: registrar/apps/api/v0/views.py [hunk 7]
```diff
@@ -195,3 +196,50 @@ def post(self, request, *args, **kwargs):
...
+                response[student_id] = request_status
+                good_status = True
+
+            seen_ids.add(student_id)
+
+        if not bad_status:
+            response_status = HTTP_200_OK
+        elif bad_status and not good_status:
+            response_status = HTTP_422_UNPROCESSABLE_ENTITY
+        else:
```
Comment (iloveagent57) [inline registrar/apps/api/v0/views.py]: Nit: it feels odd to me that you can have `bad_status` and `good_status` both be True.  Maybe there's better variable names for these?  Like `any_failures` and `any_successes`?

**Review by iloveagent57 — COMMENTED** [review]
Looks good in general, I just have some clean code suggestions.  I kinda like how you aligned those test names to status codes :)

File: registrar/apps/api/v0/views.py [hunk 8]
```diff
@@ -173,16 +174,18 @@ class MockProgramEnrollmentView(CreateAPIView, RetrieveAPIView, MockProgramSpeci
...
     permission_classes = (IsAuthenticated,)
-    serializer_class = ProgramEnrollmentRequestSerializer
 
     def post(self, request, *args, **kwargs):
+        return self.validate_and_echo_statuses(request, ProgramEnrollmentRequestSerializer, "student_key")
+
+    def patch(self, request, *args, **kwargs):
+        return self.validate_and_echo_statuses(request, ProgramEnrollmentModificationRequestSerializer, "student_id")
+
+    def validate_and_echo_statuses(self, request, serializer, student_id_name):
```
Comment (iloveagent57) [inline registrar/apps/api/v0/views.py]: tiniest of nits: `student_id_field` is clearer, IMHO.  With `student_id_name` I don't know if it's an id, or a name, or an id and a name jammed together.

Comment (jansenk) [reply registrar/apps/api/v0/views.py]: how about `student_id_field_name`?

**Review by iloveagent57 — APPROVED** [review]
Looks good, tiny suggestion only.

File: registrar/apps/api/v0/views.py (hunk 8 above)
Comment (jansenk) [reply registrar/apps/api/v0/views.py]: never mind, no longer an issue

**Review by zacharis278 — APPROVED** [review]

**Comment by zacharis278** [comment]
Just need to suppress that swagger error but otherwise lgtm

**Review by kdmccormick — APPROVED** [review]
Swagger is showing a PUT request option. Other than that, pending squashing commits, LGTM
